src/lib/server/xlat_func.c

```diff
@@ -227,100 +227,34 @@ static ssize_t xlat_length(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
  *
  */
-static ssize_t xlat_integer(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
-			    UNUSED void const *mod_inst, UNUSED void const *xlat_inst,
-			    REQUEST *request, char const *fmt)
+static xlat_action_t xlat_integer(UNUSED TALLOC_CTX *ctx, fr_cursor_t *out,
+			      REQUEST *request, UNUSED void const *xlat_inst, UNUSED void *xlat_thread_inst,
+			      fr_value_box_t **in)
+
 {
+	fr_value_box_t *vb;
 	VALUE_PAIR 	*vp;
-
-	uint64_t 	int64 = 0;	/* Needs to be initialised to zero */
-	uint32_t	int32 = 0;	/* Needs to be initialised to zero */
-
-	while (isspace((int) *fmt)) fmt++;
-
-	if ((xlat_fmt_get_vp(&vp, request, fmt) < 0) || !vp) return 0;
-
-	switch (vp->vp_type) {
-	case FR_TYPE_STRING:
-	{
-		fr_value_box_t vb;
-
-		if (fr_value_box_cast(NULL, &vb, FR_TYPE_UINT64, NULL, &vp->data) < 0) {
-			RPEDEBUG("Input string invalid");
-			return -1;
-		}
-
-		return snprintf(*out, outlen, "%" PRIu64, vb.vb_uint64);
-	}
-
-	case FR_TYPE_OCTETS:
-		if (vp->vp_length > 8) {
-			break;
-		}
-
-		if (vp->vp_length > 4) {
-			memcpy(&int64, vp->vp_octets, vp->vp_length);
-			return snprintf(*out, outlen, "%" PRIu64, htonll(int64));
-		}
-
-		memcpy(&int32, vp->vp_octets, vp->vp_length);
-		return snprintf(*out, outlen, "%i", htonl(int32));
-
-	case FR_TYPE_UINT64:
-		return snprintf(*out, outlen, "%" PRIu64, vp->vp_uint64);
+	char const *var;
 
 	/*
-	 *	IP addresses are treated specially, as parsing functions assume the value
-	 *	is bigendian and will convert it for us.
+	 *	If there's no input, there's no output
 	 */
-	case FR_TYPE_IPV4_ADDR:
-	case FR_TYPE_IPV4_PREFIX:	/* Same addr field */
-		return snprintf(*out, outlen, "%u", ntohl(vp->vp_ipv4addr));
-
-	case FR_TYPE_UINT32:
-		return snprintf(*out, outlen, "%u", vp->vp_uint32);
-
-	case FR_TYPE_DATE:
-		return snprintf(*out, outlen, "%u", vp->vp_date);
-
-	case FR_TYPE_UINT8:
-		return snprintf(*out, outlen, "%u", (unsigned int) vp->vp_uint8);
-
-	case FR_TYPE_UINT16:
-		return snprintf(*out, outlen, "%u", (unsigned int) vp->vp_uint16);
+	if (!*in) return XLAT_ACTION_DONE;
 
-	/*
-	 *	Ethernet is weird... It's network related, so it
-	 *	should be bigendian.
-	 */
-	case FR_TYPE_ETHERNET:
-		int64 = vp->vp_ether[0];
-		int64 <<= 8;
-		int64 |= vp->vp_ether[1];
-		int64 <<= 8;
-		int64 |= vp->vp_ether[2];
-		int64 <<= 8;
-		int64 |= vp->vp_ether[3];
-		int64 <<= 8;
-		int64 |= vp->vp_ether[4];
-		int64 <<= 8;
-		int64 |= vp->vp_ether[5];
-		return snprintf(*out, outlen, "%" PRIu64, int64);
-
-	case FR_TYPE_INT32:
-		return snprintf(*out, outlen, "%i", vp->vp_int32);
-
-	case FR_TYPE_IPV6_ADDR:
-	case FR_TYPE_IPV6_PREFIX:
-		return fr_snprint_uint128(*out, outlen, ntohlll(*(uint128_t const *) &vp->vp_ipv6addr));
+	var = (*in)->vb_strvalue;
+	if ((xlat_fmt_get_vp(&vp, request, var) < 0) || !vp) return XLAT_ACTION_FAIL;
 
-	default:
-		break;
+	MEM(vb = fr_value_box_alloc(ctx, vp->vp_type, NULL, false));
+	if (fr_value_box_cast(ctx, vb, vp->vp_type, NULL, &vp->data) < 0) {
```

Comment:
You appear to be casting the data to the same type as it was previously, which is just a copy operation, and then inserting a copy of the VP's data into the output list.  That's not what this function does... It's meant to return the integer representation of the data you passed in.